[ADD] pos_default_quantity: sets default orderline quantity

robinkeunen opened the pull request:
**Normal behaviour**: order line quantities are set to 0

**This module**:
- adds `pos_default_qty` field to `product.uom.categ`
- order line quantity is set to the `pos_default_qty` linked to the added product.

robinkeunen committed:
[ADD] pos_default_quantity: sets default orderline quantity

fixup! [ADD] pos_default_quantity: sets default orderline quantity

robinkeunen committed:
[ADD] pos_default_quantity: readme

chienandalu commented on pos_default_quantity/__openerp__.py:
```diff
@@ -0,0 +1,25 @@
+    "website": "www.coopiteasy.be",
+    "license": "AGPL-3",
+    "category": "Point of Sale",
+    "summary": """
```
Use description key

robinkeunen replied:
I changed it but the linting bot advises to remove description
```
pos_default_quantity/__openerp__.py:5: [C8103(manifest-deprecated-key), ] Deprecated key "description" in manifest file
```

chienandalu commented on pos_default_quantity/__openerp__.py:
```diff
@@ -0,0 +1,25 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Set Default Product Quantity in POS",
+    "version": "9.0.0.1.0",
```
9.0.1.0.0

chienandalu commented on pos_default_quantity/__openerp__.py:
```diff
@@ -0,0 +1,25 @@
+    "name": "Set Default Product Quantity in POS",
+    "version": "9.0.0.1.0",
+    "author": "Coop IT Easy SCRLfs, Odoo Community Association (OCA)",
+    "website": "www.coopiteasy.be",
```
https://github.com/OCA/pos

chienandalu commented on pos_default_quantity/models/product.py:
```diff
@@ -0,0 +1,15 @@
+class ProductUomCategory(models.Model):
+    _inherit = 'product.uom.categ'
+
+    pos_default_qty = fields.Float(
```
Why not to use something like a default_pos_uom_id and take the conversion from its conversion factor?

robinkeunen replied:
I am not sure I understand your suggestion, the goal of this PR is to allow setting a default quantity when initialising an orderline in the pos. The use case being : 
- if product uom is unit, default qty is 1
- if product uom is weight, default qty is 0

Thanks for the review btw

robinkeunen replied:
My above comment is misleading. The use case is :

- if product uom **category** is unit, default qty is 1
- if product uom **category** is weight, default qty is 0

chienandalu commented on pos_default_quantity/static/src/js/pos.js:
```diff
@@ -0,0 +1,41 @@
+
+        var orderline_prototype = models.Orderline.prototype;
+        models.Orderline = models.Orderline.extend({
+            initialize: function (attr, options) {
```
using a default_pos_uom, you'd override the get_unit function instead

chienandalu commented on pos_default_quantity/static/src/js/pos.js:
```diff
@@ -0,0 +1,41 @@
+        var models = require('point_of_sale.models');
+
+        models.load_models({
+            model: 'product.uom.categ',
```
using a default_pos_uom, you'd override the loading of product.product to get that field

chienandalu commented on pos_default_quantity/static/src/js/pos.js:
```diff
@@ -0,0 +1,41 @@
+                orderline_prototype.initialize.call(this, attr, options);
+
+                var unit = this.get_unit();
+                var category = this.pos.unit_categories_by_id[unit.category_id[0]]; // eslint-disable-line
```
Don't disable linting if not justified :)

chienandalu reviewed (changes requested):
Thanks for your contribution, @robinkeunen! Some comments and suggestions:

Also check runbot and travis warnings (they seem to be minor stuff)

chienandalu reviewed (commented):
Hi, @robinkeunen Rebase your branch to get the last CI fixes

legalsylvain commented:
Any news on that PR ?

robinkeunen commented:
@chienandalu @legalsylvain 
Tests passed on my travis, still ongoing on the oca. I'll start the migration to 12.0.

I'll be at  the OCA sprint tomorrow.

robinkeunen committed:
[FIX] pos_default_quantity: don't set default quantity on restored lines

When restoring the order from a json, previous code overwrote quantity with default quantity

ivantodorovich commented on pos_default_quantity/static/src/js/pos.js:
```diff
@@ -0,0 +1,49 @@
+*/
+
+odoo.define(
+    'pos_default_quantity.pos_default_quantity',
```
Please name the `.js` file same as your `define`.

ie:
both `pos.js` and `odoo.define("pos_default_quantity.pos")`, for example

ivantodorovich commented on pos_default_quantity/static/src/js/pos.js:
```diff
@@ -0,0 +1,49 @@
+
+        var models = require('point_of_sale.models');
+
+        models.load_models({
```
I'd add the default_qty field in the `product.uom` model instead.
If you want, it can be a related field to the `product.uom.categ`, for functional or usability reasons.

But technically speaking, the `product.uom` model is already loaded in the pos to get the `groupable` and `is_unit` fields:

https://github.com/odoo/odoo/blob/7132a15c38a3b5e5ec232404d21aacfa2ac5874e/addons/point_of_sale/static/src/js/models.js#L169-L182

So you could simply call [load_fields](https://github.com/odoo/odoo/blob/7132a15c38a3b5e5ec232404d21aacfa2ac5874e/addons/point_of_sale/static/src/js/models.js#L982) to load your extra field, and avoid loading yet another module to the pos ui:
